src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,468 @@
+  if (event_type != G_FILE_MONITOR_EVENT_CHANGES_DONE_HINT && event_type != G_FILE_MONITOR_EVENT_DELETED)
+    return;
+
+  path = g_file_get_path (file);
+  if (g_str_has_suffix (path, ".json"))
+    {
+      g_debug ("on_machines_changed: event type %i on %s", event_type, path);
+      /* change events tend to come in batches, so slightly delay the re-reading of
+       * files and sending PropertiesChanged; if we already have queued up an
+       * update, don't queue it again */
+      if (pending_updates++ == 0)
+        g_timeout_add (100, notify_properties, user_data);
+    }
+  else
+    {
+      g_debug ("on_machines_changed: ignoring event type %i on non-.json file %s", event_type, path);
+    }
+  g_free (path);
+}
+
+
+static GDBusInterfaceVTable machines_vtable = {
+  .method_call = machines_method_call,
+  .get_property = machines_get_property,
+};
+
+static GDBusPropertyInfo machines_property = {
+  -1, "Machines", MACHINES_SIG, G_DBUS_PROPERTY_INFO_FLAGS_READABLE, NULL
+};
+
+static GDBusPropertyInfo *machines_properties[] = {
+  &machines_property,
+  NULL
+};
+
+static GDBusArgInfo machines_update_filename_arg = {
+  -1, "filename", "s", NULL
+};
+
+static GDBusArgInfo machines_update_hostname_arg = {
+  -1, "hostname", "s", NULL
+};
+
+static GDBusArgInfo machines_update_info_arg = {
+  -1, "info", "a{sv}", NULL
+};
+
+static GDBusArgInfo *machines_update_args[] = {
+  &machines_update_filename_arg,
+  &machines_update_hostname_arg,
+  &machines_update_info_arg,
+  NULL
+};
+
+static GDBusMethodInfo machines_update_method = {
+  -1, "Update", machines_update_args, NULL, NULL
+};
+
+static GDBusMethodInfo *machines_methods[] = {
+  &machines_update_method,
+  NULL
+};
+
+static GDBusInterfaceInfo machines_interface = {
+  -1, "cockpit.Machines", machines_methods, NULL, machines_properties, NULL
+};
+
+void
+cockpit_dbus_machines_startup (void)
+{
+  GDBusConnection *connection;
+  GFileMonitor *machines_monitor;
+  GFile *machines_monitor_file;
+  GError *error = NULL;
+
+  connection = cockpit_dbus_internal_server ();
+  g_return_if_fail (connection != NULL);
+
+  g_dbus_connection_register_object (connection, "/machines", &machines_interface,
+                                     &machines_vtable, NULL, NULL, &error);
+
+  if (error != NULL)
+    {
+      g_critical ("couldn't register DBus cockpit.Machines object: %s", error->message);
+      g_error_free (error);
+      return;
+    }
+
+  /* watch for file changes and send D-Bus signal for it */
+  machines_monitor_file = g_file_new_for_path (machines_json_dir ());
+  machines_monitor = g_file_monitor (machines_monitor_file, G_FILE_MONITOR_NONE, NULL, &error);
+  g_object_unref (machines_monitor_file);
+  if (machines_monitor == NULL)
+    {
+      g_critical ("couldn't set up file watch: %s", error->message);
+      g_error_free (error);
+      return;
+    }
+  g_signal_connect (machines_monitor, "changed", G_CALLBACK (on_machines_changed), connection);
```

Comment:
Where is machines_monitor unreffed? Strange ... why doesn't it show up in memory leak info? I guess we're not covering cockpit-bridge startup/shutdown in our valgrind checks.

For completeness this could use a cockpit_dbus_machines_cleanup() function.